paddle/fluid/framework/custom_operator.cc

```diff
@@ -0,0 +1,285 @@
+  CustomOpMaker custom_maker(input_num);
+  info.proto_->set_type(name);
+  custom_maker(info.proto_, info.checker_);
+  PADDLE_ENFORCE_EQ(
+      info.proto_->IsInitialized(), true,
+      platform::errors::PreconditionNotMet(
+          "Fail to initialize %s's OpProto, because %s is not initialized.",
+          name, info.proto_->InitializationErrorString()));
+
+  // GradOpDescMaker
+  info.grad_op_maker_ = [](
+      const OpDesc& fwd_op, const std::unordered_set<std::string>& no_grad_set,
+      std::unordered_map<std::string, std::string>* grad_to_var,
+      const std::vector<BlockDesc*>& grad_block) {
+    DefaultGradOpMaker<paddle::framework::OpDesc, true> maker(
+        fwd_op, no_grad_set, grad_to_var, grad_block);
+    return maker();
+  };
+  // GradOpBaseMaker
+  info.dygraph_grad_op_maker_ = [](
+      const std::string& type,
+      const imperative::NameVarBaseMap& var_base_map_in,
+      const imperative::NameVarBaseMap& var_base_map_out,
+      const framework::AttributeMap& attrs,
+      const std::map<std::string, std::string>& inplace_map) {
+    DefaultGradOpMaker<paddle::imperative::OpBase, true> maker(
+        type, var_base_map_in, var_base_map_out, attrs, inplace_map);
+    return maker();
+  };
+  info.use_default_grad_op_desc_maker_ = true;
+
+  /* Grad op register */
+  OpInfo grad_info;
+
+  // Op
+  grad_info.creator_ = [](
+      const std::string& type, const VariableNameMap& inputs,
+      const VariableNameMap& outputs, const AttributeMap& attrs) {
+    return new CustomGradOperator(type, inputs, outputs, attrs);
+  };
+
+  // InferShape
+  OperatorWithKernel* grad_op =
+      dynamic_cast<OperatorWithKernel*>(grad_info.creator_(
+          std::string{}, VariableNameMap{}, VariableNameMap{}, AttributeMap{}));
+  grad_info.infer_shape_ = [grad_op](InferShapeContext* ctx) {
+    grad_op->InferShape(ctx);
+  };
+
+  // Last Step: insert
+  OpInfoMap::Instance().Insert(name, info);
+  OpInfoMap::Instance().Insert(name + "_grad", grad_info);
+}
+
+template <typename Func>
+void RegisterOperatorKernel(const std::string& name, TensorFunction* func) {
+  // 0. unwrap func
+  Func true_func = std::move(func->UnWrap<Func>());
+
+  // 1. construct kernel func
+  CustomOpKernel<float, Func> custom_kernel(true_func);
+
+  // 2. insert kernel
+  std::string library_type = "CPU";
+  std::string data_layout = "ANYLAYOUT";
+  OpKernelType key(ToDataType(std::type_index(typeid(float))),
+                   platform::CPUPlace());
+
+  VLOG(0) << "op name in kernel: " << name;
+  VLOG(0) << "op kernel key: " << key;
+  OperatorWithKernel::AllOpKernels()[name][key] =
+      [custom_kernel](const framework::ExecutionContext& ctx) {
+        VLOG(0) << "run custom kernel func.";
+        custom_kernel.Compute(ctx);
+      };
+}
+
+// load op api
+void LoadCustomOperator(const std::string& dso_name) {
+  void* handle = paddle::platform::dynload::GetOpDsoHandle(dso_name);
+
+  typedef OpFunctionMap& get_op_func_map_t();
```

Comment:
Bad name, maybe `OpFuncMapRef` to indicate reference type

Reply:
here the type is `get_op_func_map_t `, not `OpFunctionMap`